l1-contracts/contracts/bridgehub/MessageRoot.sol

```diff
@@ -120,13 +159,121 @@ contract MessageRoot is IMessageRoot, Initializable {
+            /// On Gateway we save the upgrade block number
+            v30UpgradeGatewayBlockNumber = block.number;
+        }
         _initialize();
         _disableInitializers();
     }
 
     /// @dev Initializes a contract for later use. Expected to be used in the proxy on L1, on L2 it is a system contract without a proxy.
     function initialize() external initializer {
         _initialize();
+        uint256[] memory allZKChains = BRIDGE_HUB.getAllZKChainChainIDs();
+        uint256 allZKChainsLength = allZKChains.length;
+        /// If there are no chains, that means we are using the contracts locally.
+        if (allZKChainsLength == 0) {
+            v30UpgradeGatewayBlockNumber = 1;
+        }
+    }
+
+    /// @dev The initialized used for the V30 upgrade.
+    function initializeV30Upgrade() external reinitializer(2) {
+        uint256[] memory allZKChains = BRIDGE_HUB.getAllZKChainChainIDs();
+        uint256 allZKChainsLength = allZKChains.length;
+        for (uint256 i = 0; i < allZKChainsLength; ++i) {
+            uint256 batchNumberToWrite = V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_GATEWAY;
+            if (BRIDGE_HUB.settlementLayer(allZKChains[i]) == block.chainid && block.chainid == L1_CHAIN_ID) {
+                /// If we are settling on L1.
+                batchNumberToWrite = V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_L1;
+            }
+            v30UpgradeChainBatchNumber[allZKChains[i]] = batchNumberToWrite;
+        }
+        /// If there are no chains, that means we are using the contracts locally.
+        if (allZKChainsLength == 0) {
+            v30UpgradeGatewayBlockNumber = 1;
+        }
+    }
+
+    function sendV30UpgradeBlockNumberFromGateway(uint256 _chainId, uint256) external {
+        uint256 sentBlockNumber;
+        if (_chainId != block.chainid) {
+            sentBlockNumber = v30UpgradeChainBatchNumber[_chainId];
+            require(
+                sentBlockNumber != V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_GATEWAY,
+                V30UpgradeChainBatchNumberNotSet()
+            );
+        } else {
+            sentBlockNumber = v30UpgradeGatewayBlockNumber;
+        }
+        // slither-disable-next-line unused-return
+        L2_TO_L1_MESSENGER_SYSTEM_CONTRACT.sendToL1(
+            abi.encodeCall(this.sendV30UpgradeBlockNumberFromGateway, (_chainId, sentBlockNumber))
+        );
+    }
+
+    function saveV30UpgradeGatewayBlockNumberOnL1(FinalizeL1DepositParams calldata _finalizeWithdrawalParams) external {
+        bool success = proveL1DepositParamsInclusion(_finalizeWithdrawalParams, L2_MESSAGE_ROOT_ADDR);
+        if (!success) {
+            revert InvalidProof();
+        }
+
+        require(
+            BRIDGE_HUB.whitelistedSettlementLayers(_finalizeWithdrawalParams.chainId),
+            NotWhitelistedSettlementLayer(_finalizeWithdrawalParams.chainId)
+        );
+        require(block.chainid == L1_CHAIN_ID, OnlyL1());
```

Comment:
we will anyway need to split message root into l1messageroot and l2messageroot for constructor removal purposes, please do it here as well, it will also better emphasize what is possible on L1 and what is intended for L2